staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -312,25 +312,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 		RecordTime:      w.clock.Now(),
 	}
 
+	// We can call w.store.Get() outside of a critical section,
+	// because the w.store itself is thread-safe and the only
+	// place where w.store is modified is below (via updateFunc)
+	// and these calls are serialized because reflector is processing
+	// events one-by-one.
+	previous, exists, err := w.store.Get(elem)
+	if err != nil {
+		return err
+	}
+	if exists {
+		previousElem := previous.(*storeElement)
+		wcEvent.PrevObject = previousElem.Object
+		wcEvent.PrevObjLabels = previousElem.Labels
+		wcEvent.PrevObjFields = previousElem.Fields
+	}
+
 	if err := func() error {
-		// TODO: We should consider moving this lock below after the watchCacheEvent
-		// is created. In such situation, the only problematic scenario is Replace()
-		// happening after getting object from store and before acquiring a lock.
```

Comment:
Is this concern for Replace invalid? Or do we need to consider it?

Reply:
I no longer understand this comment (even though I was the one who added this).

But the story for replace is the following:
- we're using reflector and it's reflector that is calling the Replace
- the way reflector works is that it's effectively single-threaded - it's either processing watch events (and calling Add/Modify/Delete functions which land here) or calling Replace (and then this function can't be called).

So this safe.


The only interpretation that comes to my mind is that I had in mind some generic usage of watchcache, where the functions aren't called from reflector but by some arbitrary caller in arbitrary order from multiple threads. In such case this change wouldn't be safe, but it's not how it is used.